Rework material on library components in Cabal Users Guide (#9370)
haskell/cabal: doc/cabal-package.rst

===================

The Cabal package is the unit of distribution. When installed, its
purpose is to make available one or more:

-  Haskell programs (executables); and/or

-  libraries, exposing a number of Haskell modules.

Executables can declare a library in the same package as a dependency, and
Cabal will treat them as if they were in another package that depended on the

> library. The alternative, which applied before Cabal 1.8.0.4, is that if the

Author: I agree. My proposed change (not leading with the ancient history, but not deleting it) pulled its punches.

    Build information for libraries.

    A package can include zero or more library components, being an optional
    'main' unnamed library (which is identified by having no ``name`` argument)
    and/or zero or more 'subsidiary' or 'ancilliary' named libraries (referred
    to as 'sublibraries', which are identified by their ``name`` argument). A

>     sublibrary cannot have the same name as its package.

Author: I'm happy to split into shorter sentences. My reason for referring to 'subsidiary' and 'ancillary' is to explain the (likely) origin of (and so give content to) the abbreviated term 'sublibrary' for named library components. I included both because some people preferred 'ancilliary' to 'subsidiary'. (Their meanings are very similar.)

>     sublibrary cannot have the same name as its package.

Reviewer: I'm currently wondering if "related library" is better and clearer than either. (I was originally in favor of "ancillary", but…
```
[26 01:45:36] <haskellbridge> <g​eekosaur> I considered "ancillary" but half expect people to look at me like I said "monad"…
```
(The defense rests.))

>     sublibrary cannot have the same name as its package.

Reviewer 2: Haha very prescient!

>     A sublibrary has a 'visibility' which can be 'public' or 'private' (the
>     default) - see the :pkg-field:`library:visibility` field. The 'main'
>     unnamed library is always public. A private sublibrary may be referred to as
>     an 'internal' library, because other packages cannot depend upon it.

Reviewer 2: ```suggestion
    A sublibrary may be marked 'public' or 'private' (the
    default) — see the :pkg-field:`library:visibility` field. The main
    library is always public. A private sublibrary is also known as
    an internal library. 
    Only public library components can be depended on by other packages.
```
I'm suggesting:
  - only quoting things when necessary
  - changing a dash to an em-dash
  - dropping the reference to "unnamed". I don't think that's relevant here
  - split the last sentence into two.

Author: There has been debate about whether library components should be referred to as (a) unnamed/named or (b) main/sub[sidiary]. The former is objective, but does not convey the relative 'standing' of the named library component. The latter conveys the (subjective) relative 'standing'. I avoided that choice by using "'main' unnamed library".

The quotation marks around the other words on their first introduction were also because they were being used figuratively rather than literally. An alternative would be to use code font (eg ``visibility``).

    Before Cabal 3.0, all sublibraries were internal libraries. Before Cabal 2.0
    a package could not include sublibraries.

>     A package being able to include more than one public library separates the
>     unit of distribution (the package) from the unit of buildable code (the
>     library). This is useful for Haskell projects with many libraries that are
>     distributed together as it avoids duplication and potential inconsistencies.

Reviewer 2: ```suggestion
    Being able to include more than one public library in a package allows separating the
    unit of distribution (the package) from the unit of buildable code (the
    library). This is useful for Haskell projects with many libraries that are
    distributed together as it avoids duplication and potential inconsistencies.
```
This reads slightly better to me but I feel like this is a matter of taste

Author: I agree.

    See :ref:`Sublibraries - Examples <sublibs>` for examples.

A library section should contain the following fields:

.. pkg-field:: visibility: visibility specifiers

    :since: 3.0

    :default:

>         ``private`` for sublibraries. Cannot be set for the 'main' unnamed
>         library, which is always public.

Reviewer 2: ```suggestion
        ``private`` for sublibraries. Cannot be set for the main
        library, which is always public.
```
As above, I don't think "main" needs to be quoted here and unnamed doesn't add any extra detail.

Author: As above, this was my attenpt to avoid chosing sides in the named/unamed v main/sub debate.

Author: I have now taken a different approach to this, by changing the introduction of terms.

>     Can be set to ``public`` or ``private``. If set to ``public``, other
>     components of the package that provide the library and components of other
>     packages are able to depend on the library. If set to ``private``, only
>     other components of the package that provide it are able to depend upon it;
>     components of other packages are not able to depend upon it.

Reviewer 2: ```suggestion
    Can be set to ``public`` or ``private``. If set to ``public``, then any component can depend on this library.
    If set to ``private``, then only components from this package can depend on it.
```

>     See the :pkg-field:`build-depends` field for the syntax to specify a 'main'
>     unnamed library or a sublibrary in a ``build-depends:``.

Reviewer 2: ```suggestion
    See the :pkg-field:`build-depends` field for the syntax to specify a dependency on a main library or a sublibrary.
```
As above I don't think we need to quote and called it unnamed. Also let's not say "build-depends" twice, rather let's say what it represents.

Author: I agree.

.. _sublibs:

**Sublibraries - Examples**

> An example of the use of an internal library (that is, a 'private' sublibrary)

Reviewer 2: ```suggestion
An example of the use of an internal library (that is, a private sublibrary)
```
I don't think this needs to be quoted

Author: Reflected in my subsequent changes.

> is a test suite that needs access to some internal modules in your 'main'
> unnamed library, which you do not otherwise want to expose. You could put those

Reviewer 2: ```suggestion
is a test suite that needs access to some internal modules in your main
library, which you do not otherwise want to expose. You could put those
```
As above

Author: Reflected in my subsequent changes.

    library
        exposed-modules:  Foo.Public
        build-depends:    foo:foo-internal, base >= 4.3 && < 5
        default-language: Haskell2010

    test-suite test-foo
        type:             exitcode-stdio-1.0
        main-is:          test-foo.hs
        -- NOTE: no constraints on 'foo-internal' as same-package
        --       dependencies implicitly refer to the same package instance
        build-depends:    foo:foo-internal, base
        default-language: Haskell2010

> Another example of the use of internal libraries is a package that includes one

Reviewer 2: I feel like this cuts quite a bit. Some of this overlaps with the 'visibility' field, but it's good to have this info here as well I think. Especially the explanation of shadowing seems important.

Author: On shadowing, I thought it no longer applies (from Cabal 3.4), because a sublibrary is now specified in a `build-depends:` as `package_name:library_name`. I think shaddowing was only applicable when a sublibrary could be specified as just `library_name` (which might clash with a package's name).

On the rest that is cut, it is all directly duplicative.

Author: On Cabal 3.4, see: https://cabal.readthedocs.io/en/stable/file-format-changelog.html#cabal-version-3-4:
> Dependencies to sublibraries must be specified explicitly, even for current package. For example: build-depends: mypackage:internal-lib This way you can have an internal library with the same name as some external dependency.

Author: 'Shadowing' brought back in, elsewhere.

or more executables but does not include a public library.

> Internal libraries can be used to incorporate an external dependency into a

Reviewer 2: Again the use of shadowing is key to this example. I don't think it should be cut

Author: As above, I don't think shaddowing applies, from Cabal 3.4.

Reviewer: This is talking about the file format, not the Cabal version. Note also that currently Hackage refuses versions later than 3.0 on the grounds that they may not be generally usable yet (although I think this is under review), so I'd consider 3.0 the breakpoint for talking about cabal file format versions.

Author: My reasoning was that the body of the Guide should discuss the current version of the library and the tool, and the corresponding version of the Cabal specification - because if the Guide was to discuss (at any length) the history of the specification each time it discussed a current feature of the library/tool its complexity could be off-putting to readers and, perhaps, more difficult to maintain. However, if Hackage is de facto limiting the adoption of versions of the specification, I agree it is useful for the Guide to deal with relevant history of the specification back to the point that Hackage accepts. I'll make edits to reflect to the relevant history of the specification of sublibraries in `build-depends:`, and reintroduce 'shadowing' - but not as a 'feature', given that it has subsequently been 'fixed'.

.. pkg-field:: build-depends: library list

>     Declares the *library* dependencies (from the same package or other
>     packages) required to build the current package component. See

Reviewer 2: ```suggestion
    Declares the *library* dependencies required to build the current package component. See
```
I don't think this parenthesis is adding any extra information. You are saying either X or not X, so that's everything

Author: I think it is useful to be express that a component of a package can depend on other (library) components in the same package AND library components in other packages (i.e. on 'everything').

(The fact that an executable can depend on the library in the same package is the 'Cabal 1.8.0.4' feature that was highlighted in the opening sentences of this part of the guide.)

Author: I have accepted this, as the point I wanted to make express is now expressed elsewhere.

    :pkg-field:`build-tool-depends` for declaring build-time *tool*
    dependencies. Dependencies on libraries from another package should be
    annotated with a version constraint.

    **Library Names**

>     A library is identified by the name of the package that provides it,
>     optionally followed by a colon and the library's name (available from

Reviewer 2: ```suggestion
    A library is identified by the name of its package,
    optionally followed by a colon and the library's name (available from
```

Author: This is now reflected.

>     ``cabal-version: 3.0``). If a library name is omitted, the package's 'main'
>     unnamed library will be used. To refer expressly to a package's 'main'

Reviewer 2: ```suggestion
    ``cabal-version: 3.0``). If a library name is omitted, the package's main
    library will be used. To refer expressly to a package's 'main'
```
As above.

Author: This is now reflected. The history of the Cabal specification has been moved out of the paragraph into a note.

    unnamed library, use the name of the package as the library name (for
    example, ``my-package:my-package``). More than one library from the same
    package can be specified with the shorthand syntax ``pkg:{lib1,lib2}``.

>     See the section on :pkg-section:`library` for more information about how
>     a package can specify library components for use by other components of the
>     same package or by other packages.

Reviewer 2: ```suggestion
    See :pkg-section:`library` for more information about declaring
    library components.
```
Be explicit about what this is about: declaring library components.

I think some of this is unnecessary and can be cut

Author: I have now shortened this. I preferred 'to specify' over 'to declare', because the former seems more frequent in the existing documentation. I don't think much turns on that (although editorial consistency would be nice to have).
